{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7315993-eb06-478f-86a7-e1055834b99b",
   "metadata": {},
   "outputs": [],
   "source": [
    "Q.no.1: Explain why we have to use the Exception class while creating a Custom Exception.\n",
    "Note: Here Exception class refers to the base class for all the exceptions.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Ans: \n",
    "    \n",
    "       When creating a custom exception in Python, it is recommended to inherit from \n",
    "        the Exception class or one of its subclasses.\n",
    "        Here are the reasons why using the Exception class as the base class for custom exceptions is beneficial:\n",
    "\n",
    "Consistency: Inheriting from the Exception class ensures that your custom exception follows the same pattern and structure \n",
    "as built-in exceptions in Python. This consistency makes it easier for developers \n",
    "to understand and handle your custom exception in a similar way they handle other exceptions.\n",
    "\n",
    "Exception Hierarchy: The Exception class serves as the base class for all built-in exceptions in Python. \n",
    "By inheriting from Exception, your custom exception becomes part of the exception hierarchy, \n",
    "allowing it to be categorized along with other\n",
    "exceptions. This hierarchy provides a convenient way to handle exceptions based on their types and relationships.\n",
    "\n",
    "Compatibility: Inheriting from the Exception class ensures compatibility with \n",
    "existing exception handling mechanisms and libraries.\n",
    "Since most exception handling code is designed to catch and handle exceptions derived from Exception, \n",
    "using it as the base class \n",
    "for your custom exception makes your code compatible with the existing error handling infrastructure.\n",
    "\n",
    "Standardized Behavior: The Exception class provides a set of methods and\n",
    "attributes that allow exceptions to behave consistently.\n",
    "By inheriting from Exception, your custom exception inherits these behaviors. \n",
    "For example, your custom exception can provide \n",
    "an error message through the __str__() method or additional information\n",
    "through custom attributes, just like other exceptions.\n",
    "\n",
    "Clarity and Readability: By using the Exception class as the base class, \n",
    "you make it explicit that your class is an exception. \n",
    "This improves code readability and helps other developers understand the purpose and usage of your custom exception.\n",
    "\n",
    "Overall, by inheriting from the Exception class, you ensure that your custom exception is consistent with the Python \n",
    "exception hierarchy, compatible with existing exception handling mechanisms, and follows the established conventions \n",
    "and behaviors of exceptions in Python.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Q2. Write a python program to print Python Exception Hierarchy.\n",
    "\n",
    "\n",
    "\n",
    "Ans:    \n",
    "    \n",
    "\n",
    "     A Python program that prints the Python Exception Hierarchy:\n",
    "\n",
    "\n",
    "def print_exception_hierarchy(exception_class, indent=''):\n",
    "    print(indent + exception_class.__name__)\n",
    "    if exception_class.__bases__:\n",
    "        for base_class in exception_class.__bases__:\n",
    "            print_exception_hierarchy(base_class, indent + '  ')\n",
    "\n",
    "# Start from the base Exception class\n",
    "print_exception_hierarchy(Exception)\n",
    "\n",
    "\n",
    "When you run this program, it will start from the base Exception class and recursively print the exception class names\n",
    "in the hierarchy, along with the necessary indentation to represent the inheritance relationships. \n",
    "The program uses recursion to traverse the exception hierarchy and prints each class name.\n",
    "\n",
    " An example of the output you can expect:\n",
    "\n",
    "\n",
    "Exception\n",
    "BaseException\n",
    "SystemExit\n",
    "KeyboardInterrupt\n",
    "GeneratorExit\n",
    "Exception\n",
    "StopIteration\n",
    "StopAsyncIteration\n",
    "ArithmeticError\n",
    "FloatingPointError\n",
    " OverflowError\n",
    "ZeroDivisionError\n",
    "AssertionError\n",
    "AttributeError\n",
    "BufferError\n",
    "EOFError\n",
    "ImportError\n",
    "ModuleNotFoundError\n",
    "LookupError\n",
    "IndexError\n",
    "KeyError\n",
    "MemoryError\n",
    "NameError\n",
    "UnboundLocalError\n",
    "OSError\n",
    "BlockingIOError\n",
    "ChildProcessError\n",
    "ConnectionError\n",
    "BrokenPipeError\n",
    "ConnectionAbortedError\n",
    "ConnectionRefusedError\n",
    "ConnectionResetError\n",
    "FileExistsError\n",
    "FileNotFoundError\n",
    "InterruptedError\n",
    "IsADirectoryError\n",
    " NotADirectoryError\n",
    "PermissionError\n",
    "ProcessLookupError\n",
    "TimeoutError\n",
    "ReferenceError\n",
    "RuntimeError\n",
    "RecursionError\n",
    "SyntaxError\n",
    "IndentationError\n",
    "TabError\n",
    "SystemError\n",
    "TypeError\n",
    "ValueError\n",
    "UnicodeError\n",
    "UnicodeDecodeError\n",
    "UnicodeEncodeError\n",
    "UnicodeTranslateError\n",
    "Warning\n",
    "DeprecationWarning\n",
    " PendingDeprecationWarning\n",
    "RuntimeWarning\n",
    "SyntaxWarning\n",
    "UserWarning\n",
    "FutureWarning\n",
    " ImportWarning\n",
    "UnicodeWarning\n",
    "BytesWarning\n",
    "ResourceWarning                \n",
    "\n",
    "        This program provides a visual representation of the Python Exception Hierarchy,\n",
    "showing the parent-child relationships between exception classes.\n",
    "       \n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "Q.no3:  What errors are defined in the ArithmeticError class? Explain any two with an example. \n",
    "\n",
    "\n",
    "\n",
    "Ans: \n",
    "    \n",
    "     In Python, the ArithmeticError class is the base class for exceptions that occur during arithmetic operations.\n",
    "        It represents a wide range of arithmetic errors. Some specific errors defined in the ArithmeticError class include:\n",
    "\n",
    "OverflowError: This error is raised when an arithmetic operation exceeds the maximum limit of a numeric type.\n",
    "It occurs when a calculation results in a value that is too large to be represented by the data type. For example:\n",
    "\n",
    "    import sys\n",
    "\n",
    "max_value = sys.maxsize\n",
    "\n",
    "try:\n",
    "    result = max_value + max_value\n",
    "except OverflowError as e:\n",
    "    print(\"Overflow Error:\", e)\n",
    "    \n",
    "    Output:\n",
    "\n",
    "\n",
    "Overflow Error: integer addition result too large for a C long\n",
    "In this example, we attempt to add max_value to itself, which exceeds the limit of the long data type.\n",
    "This causes an OverflowError to be raised.\n",
    "\n",
    "ZeroDivisionError: This error is raised when attempting to divide a number by zero. \n",
    "It occurs when the divisor in a division\n",
    "operation is zero. For example:\n",
    "\n",
    "    \n",
    "try:\n",
    "    result = 10 / 0\n",
    "except ZeroDivisionError as e:\n",
    "    print(\"Zero Division Error:\", e)\n",
    "\n",
    "    \n",
    "    Output:\n",
    "\n",
    "\n",
    "Zero Division Error: division by zero\n",
    "\n",
    "In this example, we try to divide the number 10 by zero, which is not a valid operation.\n",
    "This raises a ZeroDivisionError because division by zero is undefined.\n",
    "\n",
    "These are just two examples of errors defined in the ArithmeticError class. It's\n",
    "important to note that ArithmeticError itself is an abstract base class, \n",
    "so it's not usually raised directly.\n",
    "Instead, its subclasses, such as OverflowError and ZeroDivisionError, \n",
    "are typically raised when specific arithmetic errors occur.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Q.no4:  Why LookupError class is used? Explain with an example KeyError and IndexError.\n",
    "\n",
    "\n",
    "\n",
    "Ans:     \n",
    "       \n",
    "    The LookupError class is used to handle errors that occur when trying to access or look up elements in a sequence \n",
    "    (such as lists, tuples, or dictionaries) or other data structures. \n",
    "    It is a base class for more specific lookup-related error\n",
    "    classes in Python.\n",
    "\n",
    "One common subclass of LookupError is the KeyError class, \n",
    "which is raised when you try to access a dictionary using a key that \n",
    "does not exist in the dictionary. For example:\n",
    "\n",
    "my_dict = {'apple': 'red', 'banana': 'yellow', 'orange': 'orange'}\n",
    "\n",
    "try:\n",
    "    print(my_dict['grape'])\n",
    "except KeyError as e:\n",
    "    print(f\"KeyError: {e}\")\n",
    "    \n",
    "In this example, the dictionary my_dict does not have the key 'grape', so trying to access it will raise a KeyError.\n",
    "The KeyError object contains the key that caused the error, which can be accessed through the e variable. \n",
    "The output of the above code will be:\n",
    "\n",
    "KeyError: 'grape'\n",
    "\n",
    "\n",
    "Another subclass of LookupError is the IndexError class, which is raised when you try to access a sequence\n",
    "(like a list or tuple) using an index that is out of range. For example:\n",
    "\n",
    "\n",
    "my_list = [1, 2, 3]\n",
    "\n",
    "try:\n",
    "    print(my_list[5])\n",
    "except IndexError as e:\n",
    "    print(f\"IndexError: {e}\")\n",
    "    \n",
    "In this example, the list my_list has only three elements, so trying to access index 5 (which is out of range)\n",
    "will raise an IndexError. The IndexError object contains the index that caused the error, which can be accessed through\n",
    "the e variable. \n",
    "The output of the above code will be:\n",
    "\n",
    "IndexError: list index out of range\n",
    "\n",
    "Both KeyError and IndexError are specific types of LookupError that allow you to handle these lookup-related errors\n",
    "in a more targeted manner. By using the appropriate subclass of LookupError, you can catch and handle these exceptions\n",
    "separately and provide specific error handling logic based on the nature of the error.\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "Q.no.5: Explain ImportError. What is ModuleNotFoundError?    \n",
    "    \n",
    "    \n",
    "    \n",
    "Ans:    In Python, an ImportError is raised when there is a problem importing a module or when a module cannot be found. \n",
    "It is a built-in exception that indicates an error related to importing and using modules in your Python code.\n",
    "\n",
    "The ImportError can occur due to various reasons, such as:\n",
    "\n",
    "The module you are trying to import does not exist or is not installed.\n",
    "There is an issue with the module's file or its dependencies.\n",
    "There is a syntax error or other problem within the module itself.\n",
    "\n",
    " An example that demonstrates an ImportError:\n",
    "\n",
    "\n",
    "try:\n",
    "    import non_existing_module\n",
    "except ImportError as e:\n",
    "    print(f\"ImportError: {e}\")\n",
    "    \n",
    "In this example, we are trying to import a module named non_existing_module which doesn't exist.\n",
    "As a result, an ImportError is raised. The error message will indicate that the module was not found, like this:\n",
    "\n",
    "\n",
    "\n",
    "ImportError: No module named 'non_existing_module'\n",
    "Another related exception is the ModuleNotFoundError, which is a subclass of ImportError. It specifically indicates that a\n",
    "module could not be found. Starting from Python 3.6, ModuleNotFoundError is raised instead of the more general ImportError \n",
    "when a module cannot be found.\n",
    "\n",
    " An example that raises a ModuleNotFoundError:\n",
    "\n",
    "\n",
    "try:\n",
    "    import non_existing_module\n",
    "except ModuleNotFoundError as e:\n",
    "    print(f\"ModuleNotFoundError: {e}\")\n",
    "    \n",
    "This code will produce the same error message as the previous example:\n",
    "\n",
    "ModuleNotFoundError: No module named 'non_existing_module'\n",
    "\n",
    "The distinction between ImportError and ModuleNotFoundError is mainly a matter of specificity. \n",
    "ModuleNotFoundError provides a more precise and informative error message when a module cannot be found,\n",
    "while ImportError is a more generic exception\n",
    "that can also be raised for other import-related issues.\n",
    "\n",
    "It important to handle these exceptions appropriately in your code, as they can help you identify and resolve problems\n",
    "with module imports, ensuring the smooth execution of your Python programs.    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "Q.no6: List down some best practices for exception handling in python.   \n",
    "\n",
    "\n",
    "\n",
    "Ans:   \n",
    "      Exception handling is an essential part of writing robust and reliable code in Python.\n",
    "        Here are some best practices for exception handling:\n",
    "\n",
    "Be specific with exceptions: Catch specific exceptions rather than using a generic except clause. \n",
    "This allows you to handle different exceptions differently and provides more precise error messages. \n",
    "Use multiple except blocks to handle specific exceptions.\n",
    "\n",
    "Use try-except-else: Use the try-except-else block when you have code that should only run if no exceptions occur. \n",
    "This helps separate the code that may raise an exception from the code that should run when everything is successful.\n",
    "\n",
    "Use finally for cleanup: The finally block allows you to define code that runs \n",
    "regardless of whether an exception occurs or not.\n",
    "It is useful for releasing resources, closing files, or cleaning up connections.\n",
    "\n",
    "Avoid bare except: Avoid using bare except clauses as they catch all exceptions, including system-exiting exceptions like\n",
    "SystemExit and KeyboardInterrupt. Catching these exceptions can make it difficult to stop the program or debug issues.\n",
    "\n",
    "Log or report exceptions: Always log or report exceptions instead of silently ignoring them. This helps in diagnosing and\n",
    "fixing issues. Use the logging module to log exceptions with relevant information.\n",
    "\n",
    "Reraise exceptions when necessary: If you catch an exception but cannot handle it properly, it is often better to reraise\n",
    "it using the raise statement. This preserves the original exception's \n",
    "stack trace and allows higher-level code to handle it.\n",
    "\n",
    "Handle expected exceptions: Identify exceptions that can be reasonably expected in your code and handle them explicitly. \n",
    "This makes the code more predictable and easier to maintain.\n",
    "\n",
    "Use context managers: Use context managers (e.g., the with statement) for resources that need to be properly managed, \n",
    "such as file operations or database connections. Context managers handle exceptions and resource cleanup automatically.\n",
    "\n",
    "Use custom exception classes: Define custom exception classes for specific error conditions in your code. This helps \n",
    "in better categorization of errors and makes the code more readable.\n",
    "\n",
    "Test exception handling: Write tests that specifically target exception handling scenarios to ensure your code behaves\n",
    "as expected when exceptions are raised. Test both expected and unexpected exceptions.\n",
    "\n",
    "Remember, the specific best practices may vary depending on the context and requirements of your code. \n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
